[Lens] Add gauges visualization

mbondyra opened the pull request:
## Summary

Leftovers: https://github.com/elastic/kibana/issues/121376

<img width="932" alt="Screenshot 2021-11-26 at 15 55 03" src="https://user-images.githubusercontent.com/4283304/143598890-4a7074d8-847d-4980-ac49-c8b93c06c495.png">

Fixes https://github.com/elastic/kibana/issues/89859 

### Default settings: 
- chart displays when having a metric there. Minimum value defaults to 0 (or metric - 10 when less than zero) and maximum value defaults to the first 'nicely rounded number':
- the title is taken from the metric dimension label. 
- the subtitle is not displayed. The colors are not displayed.
- the numbers are rounded to whole unless (max-min<5), then it's 0.0.
- the suggestion with the opposite gauge is only shown when gauge is displayed (so when having vertical active, we show horizontal)

<img width="965" alt="Screenshot 2021-11-25 at 16 54 42" src="https://user-images.githubusercontent.com/4283304/143471806-c41ffb3d-419f-4579-a328-b48911409eca.png">

### Visualization settings
<img width="761" alt="Screenshot 2021-11-25 at 16 58 16" src="https://user-images.githubusercontent.com/4283304/143472330-4656bce9-835e-422e-bc21-fbb541c358d1.png">


### Dimensions settings

#### Metric
- formatting
- band colors
- color palette - color palette is by default transparentized, but user can choose not transparent colors if they want to.
<img width="766" alt="Screenshot 2021-11-25 at 17 00 06" src="https://user-images.githubusercontent.com/4283304/143472560-30ec5bf8-a241-40ab-861e-2225ebd92576.png">
- ticks and color bands are not aligned unless user click on the setting in Visualization settings

<img width="950" alt="Screenshot 2021-11-25 at 17 01 49" src="https://user-images.githubusercontent.com/4283304/143472801-ef77ba19-d8c9-449b-8f2d-a1b7dda21682.png">
<img width="746" alt="Screenshot 2021-11-25 at 17 02 10" src="https://user-images.githubusercontent.com/4283304/143472855-5f78ffa1-5498-4e25-b5f1-108a5b70a304.png">

<img width="941" alt="Screenshot 2021-11-25 at 17 04 06" src="https://user-images.githubusercontent.com/4283304/143473132-09f51d32-d216-4756-9780-b9deac84abe8.png">

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

dej611 reviewed (commented):
When not bands are configured and the current value overflows the maximum value set, a green band appears somehow.

Easiest repro steps:
* Create a gauge for the last 24h and set the maximum value proposed.
* Now change the interval to last 7 days

<img width="1124" alt="Screenshot 2021-11-29 at 16 03 16" src="https://user-images.githubusercontent.com/924948/143891658-43b8484c-da91-476e-b3c6-281353f218a4.png">
![gauge_green](https://user-images.githubusercontent.com/924948/143891780-4ac8c4dc-9e8f-44aa-be20-dd1277f7515f.gif)

dej611 reviewed (commented):
Have few questions when testing it locally:

* can we add a tooltip here to suggest/explain why is disabled and how to enable ticks on color bands?

<img width="460" alt="Screenshot 2021-11-29 at 16 06 43" src="https://user-images.githubusercontent.com/924948/143892229-1f391f16-3c77-484e-affe-ab95eeba41b6.png">

* in some case it seems there is a rounding issue? 300 vs 303?
  * can reproduce this only with bands on.

<img width="980" alt="Screenshot 2021-11-29 at 16 14 15" src="https://user-images.githubusercontent.com/924948/143893514-e05ef008-ae8a-471f-9f13-ca72500c5755.png">

* In case of no title/subtitle is it possible to make a better use of chart space?

<img width="759" alt="Screenshot 2021-11-29 at 16 17 40" src="https://user-images.githubusercontent.com/924948/143894433-df0f1b18-eb8c-448d-b33f-e82530827d66.png">

* When no bands are defined there's a uniform greenish background set

<img width="1124" alt="Screenshot 2021-11-29 at 16 19 26 (2)" src="https://user-images.githubusercontent.com/924948/143894535-06d8c347-bd31-4a87-936a-af3717b6b4eb.png">

* metric/goal auto static seems to have some problems with negative values:

![goal_positive](https://user-images.githubusercontent.com/924948/143896239-c9a80dbb-36a6-48a5-9881-3d63646bb39e.gif)

![goal_negative](https://user-images.githubusercontent.com/924948/143896292-50d19093-bcd2-4be5-8644-c66856ee3df0.gif)

  * I think that the magic metric static value is a bit confusing, as it can have side effects on max value

![automagic_metric](https://user-images.githubusercontent.com/924948/143896488-1f625b2a-323a-4bc3-987f-4e1fd6f78898.gif)

dej611 commented on x-pack/plugins/lens/public/visualizations/gauge/utils.ts:
```diff
@@ -0,0 +1,113 @@
+  }
+  const minValue = getMinValue(row, state);
+  const maxValue = getMaxValue(row, state);
+  return Math.round((minValue + maxValue) * 0.6);
```
I think that formula is not robust for negative values.
This should work better:

```suggestion
  return Math.round((maxValue - minValue) * 0.6 + minValue);
```

dej611 commented on x-pack/plugins/lens/public/visualizations/gauge/utils.ts:
```diff
@@ -0,0 +1,113 @@
+  }
+  const minValue = getMinValue(row, state);
+  const maxValue = getMaxValue(row, state);
+  return Math.round((minValue + maxValue) * 0.8);
```
Same as above:

```suggestion
  return Math.round((maxValue - minValue) * 0.8 + minValue);
```

ghudgins commented:
minor things / definitely nits:
 - Title discussion from @gvnmagni - linking up https://github.com/elastic/elastic-charts/issues/1507
 - Vis does not use horizontal space efficiently https://github.com/elastic/elastic-charts/issues/1508
 - the vis is off center right before the mobile breakpoint. note the label gets cropped....probably related to above 
 
![image](https://user-images.githubusercontent.com/5465514/143895061-f3cc0fb7-d7d8-4f40-8961-95bcab70d51a.png)

mbondyra commented:
> When not bands are configured and the current value overflows the maximum value set, a green band appears somehow.

**Solved - now if metric value overflows, we display a warning and set it to maximum**

> Can we add a tooltip here to suggest/explain why is disabled and how to enable ticks on color bands?

**Great idea, I need help with copy cc @MichaelMarcialis **
<img width="228" alt="Screenshot 2021-11-30 at 13 20 02" src="https://user-images.githubusercontent.com/4283304/144046322-01e2af30-5b14-4f79-89ac-415fa01bf709.png">

> In some case it seems there is a rounding issue? 300 vs 303?

**Fixed, it was caused by allowing the color bands to overflow maximum and the behavior changed now**
 
> In case of no title/subtitle is it possible to make a better use of chart space?

**There’s an issue in elastic-charts, Graham mentions it above**

> When no bands are defined there's a uniform greenish background set

**Removed, for some reason I thought we want it but checked the design and you’re right, I should have removed it;)** 

> Metric/goal auto static seems to have some problems with negative values:
> I think that the magic metric static value is a bit confusing, as it can have side effects on max value

**Solved!**

MichaelMarcialis reviewed (commented):
This is looking great so far, @mbondyra. Well done! Leaving you a few comments below for my review:

- In the toolbar's appearance menu, the selector in the title and subtitle rows is cutting off the last character of the word "Custom" when it is selected. Anything we can do to fix that?

![image](https://user-images.githubusercontent.com/3884767/144140785-306d57ed-9637-4484-b0aa-488b8a6f9fb3.png)

- Since we've recently simplified the tick options to be a single boolean to apply ticks on color bands (rather than the previous button group with none/auto/color bands options), I'm wondering if we should move this option to appear within the metric dimension configuration. It could conditionally appear when color bands are turned on (like the color preview and edit button). It could also default to being on when color bands are enabled (assuming this is something we think most users will want enabled when color bands are enabled). Thoughts?

- When creating a new gauge visualization (i.e. no dimensions populated yet), both minimum and maximum value dimension drag zones are given the new suggested value treatment by default (using 0 and 100 values respectively). Should we be showing these suggested values by default? I worry it may confuse users. My expectation was that we only wanted to show these suggested values after the user has populated the metric dimension.

![image](https://user-images.githubusercontent.com/3884767/144141381-be3b7aa5-985d-4b5b-904e-da5768f2ff41.png)

- This is a broader consistency question regarding our shift to using these new suggested values when adding fields to a gauge visualization. When a field is added to the workspace for a gauge visualization, Lens fully populates the metric value dimension and provides suggested values for the minimum and maximum value dimensions. Conversely, when dragging a field to the workspace for any other visualization type, Lens fully populates all required dimensions (i.e. doesn't use this new suggestion system). Do we feel that the gauge visualization type is unique enough to warrant this inconsistency? Or should this notion of suggested dimension values carry over to other visualization types? Or is it simply a stylistic matter where we should style these "suggested" dimension values the same way we would a fully populated dimension value to create visual consistency?

- Both minimum and maximum value dimensions are being given an "optional" label append. I assume we're doing this because of the new suggested value functionality, but I wonder if suggesting values truly means the dimension is optional. My first instinct is to say these are both still required dimensions. Lens is just doing it for you in those suggested instances. Thoughts on removing the "optional" label append for these two dimensions?

- Out of curiosity, how are the what are the current formulas being used for the suggested maximum values and default goal values?

- Currently an empty prompt populates the workspace when either 1) the minimum and maximum values are equal, or 2) the maximum value is greater than the minimum value. In addition to these empty prompts, I'm wondering if we should show the minimum value and maximum value dimension items in an error state (red text, alert icon) to better emphasize the problem areas. Thoughts?

- If the goal value is less than the minimum value or greater than the maximum value, the goal is rendered at the very beginning or end of the visualization respectively. Should we not be simply hiding the goal from the visualization, since it is set to appear outside of the gauge range in those circumstances? I feel like the current rendering of putting it at the beginning or end may cause users to think the goal value has changed.

![image](https://user-images.githubusercontent.com/3884767/144142962-8cc5b419-9ce3-4513-b9cc-1fd4a4a83d2c.png)

- In the case where a minimum value is greater than the metric value and the user also has a number-based (absolute) color range applied, the rendered visualization appears to begin at the second color range value, rather than the defined minimum value. Is this intentional? If so, why? I imagine it would cause less confusion on the users part to simply honor the minimum value they define, regardless of what their color range is.

![image](https://user-images.githubusercontent.com/3884767/144143169-f89945ea-ab33-4e6b-bf73-61c03ff0f2f0.png)

- If the user has made any color customizations, turning band colors off and on again causes those customizations to be lost and the default grayscale palette is reapplied. Any way to store those the last customizations made so they are lost when the user toggles color bands on and off?

mbondyra commented:
Thanks so much for a good feedback Michael!!! I'll divide my responses to @MichaelMarcialis  *DONE* and *to be discussed* 

### DONE
> * In the toolbar's appearance menu, the selector in the title and subtitle rows is cutting off the last character of the word "Custom" when it is selected. Anything we can do to fix that?

✅
<img width="415" alt="Screenshot 2021-12-01 at 09 18 51" src="https://user-images.githubusercontent.com/4283304/144197355-795a66b8-cdc9-485f-9bb3-5e093304988b.png">

> * When creating a new gauge visualization (i.e. no dimensions populated yet), both minimum and maximum value dimension drag zones are given the new suggested value treatment by default (using 0 and 100 values respectively). 

✅

> * If the goal value is less than the minimum value or greater than the maximum value, the goal is rendered at the very beginning or end of the visualization respectively. Should we not be simply hiding the goal from the visualization, since it is set to appear outside of the gauge range in those circumstances? I feel like the current rendering of putting it at the beginning or end may cause users to think the goal value has changed.

✅

> * In the case where a minimum value is greater than the metric value and the user also has a number-based (absolute) color range applied, the rendered visualization appears to begin at the second color range value, rather than the defined minimum value. Is this intentional? If so, why? I imagine it would cause less confusion on the users part to simply honor the minimum value they define, regardless of what their color range is.

✅ It's a bug, fixed!

> * If the user has made any color customizations, turning band colors off and on again causes those customizations to be lost and the default grayscale palette is reapplied. Any way to store those the last customizations made so they are lost when the user toggles color bands on and off?

✅ That's a very good point and it's a problem with any of the custom coloring in Lens (metric, table, gauge) - I corrected it here for gauges but maybe we want to align this behavior across others too? If so, do we need an extra `reset` button? (I was using turn off/turn on button to reset my bands when testing but maybe it's not really a usecase for users)

> * Since we've recently simplified the tick options to be a single boolean to apply ticks on color bands (rather than the previous button group with none/auto/color bands options), I'm wondering if we should move this option to appear within the metric dimension configuration. It could conditionally appear when color bands are turned on (like the color preview and edit button). It could also default to being on when color bands are enabled (assuming this is something we think most users will want enabled when color bands are enabled). Thoughts?

✅ Done!
<img width="304" alt="Screenshot 2021-12-01 at 13 20 59" src="https://user-images.githubusercontent.com/4283304/144233507-1c907d08-4500-44de-9f48-33e4b45562a1.png">

### NEEDS RESEARCH/FEEDBACK

> * Both minimum and maximum value dimensions are being given an "optional" label append. I assume we're doing this because of the new suggested value functionality, but I wonder if suggesting values truly means the dimension is optional. My first instinct is to say these are both still required dimensions. Lens is just doing it for you in those suggested instances. Thoughts on removing the "optional" label append for these two dimensions?

At the moment we have two states: not required and required. When we have not required, we get an extra label 'optional', and for required validation is displayed (on the screenshot). 
Could you confirm if right now we don't want any of these? 

<img width="287" alt="Screenshot 2021-12-01 at 13 21 45" src="https://user-images.githubusercontent.com/4283304/144233726-3902079c-062c-4c3e-bdd9-bc484be2d4df.png">

> * Currently an empty prompt populates the workspace when either 1) the minimum and maximum values are equal, or 2) the maximum value is greater than the minimum value. In addition to these empty prompts, I'm wondering if we should show the minimum value and maximum value dimension items in an error state (red text, alert icon) to better emphasize the problem areas. Thoughts?

This requires bigger effort than one would think but I'll spend some time today to try to achieve it. 

> * Out of curiosity, how are the what are the current formulas being used for the suggested maximum values and default goal values?

We can change these formulas because I've came up with them by myself without giving a lot of thought:

Minimum Value
Fallbacks to 0
Special cases:  if metric or max exists and is <0, it equals to smaller of metric or max - 10

Max value
Fallbacks to 100.
Special cases:
If min/metric/goal exists, it checks the range they create (for example, min =0 and metric = 89) and takes the next nicely rounded value using scaleLinear().domain([minValue, biggerValue]).nice()


Metric
Equals to 60% of the [min, max] range

Goal value
Equals to 80% of the [min, max] range



> * This is a broader consistency question regarding our shift to using these new suggested values when adding fields to a gauge visualization. When a field is added to the workspace for a gauge visualization, Lens fully populates the metric value dimension and provides suggested values for the minimum and maximum value dimensions. Conversely, when dragging a field to the workspace for any other visualization type, Lens fully populates all required dimensions (i.e. doesn't use this new suggestion system). Do we feel that the gauge visualization type is unique enough to warrant this inconsistency? Or should this notion of suggested dimension values carry over to other visualization types? Or is it simply a stylistic matter where we should style these "suggested" dimension values the same way we would a fully populated dimension value to create visual consistency?

I don’t have a strong opinion here. The only thing we cannot do for now is to fully populate the dimension with the real dimension, but we can style it to be viewed as real. I think it’s a good solution but there are couple of things - for example, in this case we shouldn’t have the `x` closing button. That might be confusing for the user because they wouldn’t know if they can drop a field there.

MichaelMarcialis commented:
Thanks, @mbondyra! My responses are below.

> > * Since we've recently simplified the tick options to be a single boolean to apply ticks on color bands (rather than the previous button group with none/auto/color bands options), I'm wondering if we should move this option to appear within the metric dimension configuration. It could conditionally appear when color bands are turned on (like the color preview and edit button). It could also default to being on when color bands are enabled (assuming this is something we think most users will want enabled when color bands are enabled). Thoughts?
> 
> ✅ Done, could you review copy and order? 
>
> <img alt="Screenshot 2021-12-01 at 13 20 59" width="304" src="https://user-images.githubusercontent.com/4283304/144233507-1c907d08-4500-44de-9f48-33e4b45562a1.png">

Thanks for making this change! Can shorten the label to read `Ticks on bands` and move the form row to appear after the color preview and edit row?

<hr/>

> > * In the toolbar's appearance menu, the selector in the title and subtitle rows is cutting off the last character of the word "Custom" when it is selected. Anything we can do to fix that?
> 
> ✅ I made it wider, is it ok? 
>
> <img alt="Screenshot 2021-12-01 at 09 18 51" width="415" src="https://user-images.githubusercontent.com/4283304/144197355-795a66b8-cdc9-485f-9bb3-5e093304988b.png">

Thanks, that works!

<hr/>

> > * If the user has made any color customizations, turning band colors off and on again causes those customizations to be lost and the default grayscale palette is reapplied. Any way to store those the last customizations made so they are lost when the user toggles color bands on and off?
> 
> ✅ That's a very good point and it's a problem with any of the custom coloring in Lens (metric, table, gauge) - I corrected it here for gauges but maybe we want to align this behavior across others too? If so, do we need an extra `reset` button? (I was using turn off/turn on button to reset my bands when testing but maybe it's not really a usecase for users)

Thanks for making this change. I agree, aligning this behavior across all other visualizations that use the color range UI would be great. I leave it up to you whether it should be part of this PR or split out as a separate issue. As for the reset button, I don't think it's necessary in this case, as applying one of the offered prefabricated color palettes is essentially resetting color and values.

<hr/>

> > * Both minimum and maximum value dimensions are being given an "optional" label append. I assume we're doing this because of the new suggested value functionality, but I wonder if suggesting values truly means the dimension is optional. My first instinct is to say these are both still required dimensions. Lens is just doing it for you in those suggested instances. Thoughts on removing the "optional" label append for these two dimensions?
> 
> ✅ Before this change we had two states: not required and required. When we have not required, we get an extra label 'optional', and for required validation is displayed (on the screenshot). Could you confirm if right now we don't want to display any of these?
> 
> <img alt="Screenshot 2021-12-01 at 13 21 45" width="287" src="https://user-images.githubusercontent.com/4283304/144233726-3902079c-062c-4c3e-bdd9-bc484be2d4df.png">

Correct, when a suggested value is in play, we shouldn't be showing the "Requires field" message or styling the header in red.

<hr/>

> > * Out of curiosity, how are the what are the current formulas being used for the suggested maximum values and default goal values?
> 
> We can change these formulas because I've came up with them by myself without giving a lot of thought:
> 
> **Minimum Value** Fallbacks to 0 Special cases: if metric or max exists and is <0, it equals to smaller of metric or max - 10
> 
> **Max value** Fallbacks to 100. Special cases: If min/metric/goal exists, it checks the range they create (for example, min =0 and metric = 89) and takes the next nicely rounded value using d3 function `scaleLinear().domain([minValue, biggerValue]).nice()`
> 
> **Metric** Equals to 60% of the [min, max] range
> 
> **Goal value** Equals to 80% of the [min, max] range

This largely makes sense to me. The only question I have is regarding the metric and goal value defaults. I believe we're currently putting reference lines (aka thresholds) at 75% of the range. Should we be doing the same here with the goal value default for consistency? And should the metric value be 50% of the range by default, to avoid attaching any negative or positive connotation to it by default? CCing @ghudgins and @gvnmagni for their thoughts as well.

<hr/>

> > * This is a broader consistency question regarding our shift to using these new suggested values when adding fields to a gauge visualization. When a field is added to the workspace for a gauge visualization, Lens fully populates the metric value dimension and provides suggested values for the minimum and maximum value dimensions. Conversely, when dragging a field to the workspace for any other visualization type, Lens fully populates all required dimensions (i.e. doesn't use this new suggestion system). Do we feel that the gauge visualization type is unique enough to warrant this inconsistency? Or should this notion of suggested dimension values carry over to other visualization types? Or is it simply a stylistic matter where we should style these "suggested" dimension values the same way we would a fully populated dimension value to create visual consistency?
> 
> I don’t have a strong opinion here. The only thing we cannot do for now is to fully populate the dimension with the real dimension, but we can style it to be viewed as real. I think it’s a good solution but there are couple of things to take into account - for example, in this case we shouldn’t have the `x` closing button. That might be confusing for the user because they wouldn’t know if they can drop a field there.

Yeah, I have the same exact concern with styling the suggested results (i.e. concealing the drop zone). I just want to make sure we're comfortable with the inconsistency because we feel that gauge suggestions are different enough from fully populating all required dimensions as we do in other visualization types. @ghudgins, are you comfortable with the current approach?

<hr/>

> > The reason I originally suggested place the "Experimental" badge after the alert icon for each visualization option was because of the appearance of an enter key symbol badge that appears when using keyboard controls in the menu.
> 
> Ahhh, that's a good reason. Leaving it up to @ghudgins, I'd probably revert to the old order.

Yeah, if you don't mind, let's revert to the order we had originally (warning icon first, experimental badge last). I know there's a smaller percentage of keyboard-only users, but the sandwiching of the warning icon looks worse, in my opinion.

ghudgins commented:
>Should we be doing the same here with the goal value default for consistency? And should the metric value be 50% of the range by default, to avoid attaching any negative or positive connotation to it by default? CCing @ghudgins and @gvnmagni for their thoughts as well.

Consistency is good. I don't think a keen eye would really notice but happy to align. No preference between 75 & 80

>@ghudgins, are you comfortable with the current [suggestion] approach [where gauge is special]?

I'm okay with them being different for now. I recalled we decided to do this because it was complex to make them "real" dimensions with their X. I think the UX is acceptable for the first release...and keeps it simple. We can always solve this. I don't think of this as a pattern to apply in other places but maybe it'll grow on us? I didn't mind it when playing with it.

>Ahhh, that's a good reason. Leaving it up to @ghudgins, I'd probably revert to the old order.

I'm outnumbered 2 to 1! Personally don't mind either way...I do prefer the alignment....but maybe that's because I love me a sandwich `:homer:`

mbondyra commented:
I think I addressed all the feedback from @dej611, @MichaelMarcialis and @ghudgins except from highlighting dimension buttons in red when visualization is in error state. 

When digging, I realized that this change is big and quite risky so I prefer to have a separate PR for it for two reasons - to measure the impact on performance in separation and, if we would have to revert it, to not revert the whole gauges PR. I submitted an issue with [leftovers](https://github.com/elastic/kibana/issues/120169).

stratoula commented on x-pack/plugins/lens/common/expressions/gauge_chart/gauge_chart.ts:
```diff
@@ -0,0 +1,134 @@
+    subtitle: {
+      types: ['string'],
+      help: i18n.translate('xpack.lens.gaugeChart.config.subtitle.help', {
+        defaultMessage: 'Specifies the Subtitle of the gauge chart',
```
nit: subtitle doesn't need to be in camel case

stratoula commented on x-pack/plugins/lens/public/visualizations/gauge/chart_component.tsx:
```diff
@@ -0,0 +1,233 @@
+    <Chart>
+      <Settings debugState={window._echDebugStateFlag ?? false} theme={chartTheme} />
+      <Goal
+        id="spec_1"
```
nit: I would change it to something else, i.e. `goal`

stratoula reviewed (commented):
This is amazing Marta!! Very nicely done :) I have some minor questions and 1 bug (the only one that I found)
I went through all the comments but forgive me if the questions below have been discussed and I missed that!
- I see that if I add a new color stop, the ticks are overlapping. Is it something we can control or it is how EC work?
![image](https://user-images.githubusercontent.com/17003240/144408847-8cbec9bb-9ec4-47f1-8e40-c9f71b9c9bee.png)
- I see that 100% color stop is not applied but 99% is. This happens because 100% is actually the max value. Correct?
![image](https://user-images.githubusercontent.com/17003240/144410531-f7c451aa-f039-4fb3-911d-e109c9b3bdc0.png)
- For the labels (major and minor) why we haven't followed the same UI as the XY?
![image](https://user-images.githubusercontent.com/17003240/144411563-900d9495-9436-4d1c-b0ab-5cb37e186a9a.png)
- If I have an initial configuration like this (I am choosing number range and change slightly a number)
![image](https://user-images.githubusercontent.com/17003240/144412829-83d25487-3165-4f7c-9782-9c5b03669ea3.png)
then I drag and drop a field to the min value and go back to the palette to change it to percent I see 
![image](https://user-images.githubusercontent.com/17003240/144415625-bbff5996-697f-40e6-b639-06bbca6df90a.png)


This also happens if for example I have this conf
![image](https://user-images.githubusercontent.com/17003240/144413432-5e9c10d7-d578-4f25-97f4-6b7c317ad514.png)
and go to the palette and add another color stop

mbondyra commented:
Thanks for your review Stratoula, here's my responses: 

> I see that if I add a new color stop, the ticks are overlapping. Is it something we can control or it is how EC work?

* EC work :) 

> I see that 100% color stop is not applied but 99% is. This happens because 100% is actually the max value. Correct?

* Yes!

> For the labels (major and minor) why we haven't followed the same UI as the XY?

* We are going to follow with the new design pattern for all the other places, here's an abstract from Figma:
<img width="849" alt="Screenshot 2021-12-02 at 14 24 49" src="https://user-images.githubusercontent.com/4283304/144430537-c19921c9-9d57-4894-b0d5-4238172bbc73.png">


Thanks a lot of the thorough testing, I introduced this bug while fixing another one yesterday. Fixed!

mbondyra commented:
@elasticmachine merge upstream

MichaelMarcialis reviewed (commented):
Thanks for making those requested changes so quickly, @mbondyra! This looks wonderful. In my last test, I found two (hopefully small) items that I comment on below. Once these are addressed, I think I'm good to approve.

- I noticed that the layer panels for gauge-type visualizations don't have a layer-level chart selector. Should we have a selector there that allows users to switch between compatible chart types (i.e. switch between horizontal and vertical gauges)?

![image](https://user-images.githubusercontent.com/3884767/144476218-0c1a4614-ff36-485a-b1e4-a45a01c7ed8d.png)

- For the band color configurations, do we need to include the color continuity options (at least until the new color range enhancements are integrated into Lens)? Right now it appears to the user cannot change from the default "above range" behavior.

![image](https://user-images.githubusercontent.com/3884767/144476624-8aeb73c8-f9f9-48b4-b531-f8fee852b94d.png)

mbondyra commented:
Discussed offline: 
1. let’s keep it as you have it for now (i.e. only having the chart selector on visualization layer panels when it’s a chart type that supports multi-layer) and revisit whenever we get around to addressing the chart pickers.
2. I added an item in the leftovers issue. It would be much easier to achieve something meaningful with the new color stops UI so I'll wait till this PR gets merged. If for some reason we would have problems with delivering it for 8.1, I am gonna add old color continuity options for 8.1.

dej611 reviewed (approved):
Tested the feature locally on Safari and could not spot bugs 👍 .

I have still few questions for things which are not 100% clear to me:

* Why having a static value by default for a metric? Maybe there's a use case I'm missing here.

* What should happen for array values and last_value?
  * I've tried to use the ecommerce dataset with the `price` field and `last_value` and got "no results".
  
<img width="1123" alt="Screenshot 2021-12-03 at 14 33 20" src="https://user-images.githubusercontent.com/924948/144611436-45e56fdb-1cc9-4cf2-884e-22f5d7ea7175.png">

  * I thought that maybe there were no values, so I've added a filtered metric there, still no results:
 
  <img width="1125" alt="Screenshot 2021-12-03 at 14 33 48" src="https://user-images.githubusercontent.com/924948/144611523-400374dc-4c87-417a-b042-4015f3407712.png">
    
  * Maybe it makes sense to not show anything as "multiple" values are in the field, but a warning might be helpful here. Also, suggestions should be filtered if leads to the same results, I think.


* Ticks on bands could propose a help icon there to explain what that means.
  * sometimes enabling the feature doesn't do much as there's no misalignment yet between ticks and bands.
  
  ![gauge_follow_up_help](https://user-images.githubusercontent.com/924948/144612015-a7e8c9f8-38b0-47a4-9177-2812a5c85dd8.gif)

* While I like the "suggested" mechanism here for min/max, I think that it is too easy to have a gauge with no defined min/max on a dashboard. I do not see this type of chart particularly useful, unless has been already defined a use case for that.

  ![gauge_suggested_minmax](https://user-images.githubusercontent.com/924948/144612602-de9cb7a5-5ae1-4677-a488-8089803dd7a2.gif)

    * If not preventing the user to save that type of gauge, can we, at least show some warning on save about that? Even better would be a warning on the Lens panel in the dashboard explaining the problem.

MichaelMarcialis reviewed (approved):
Per our conversation yesterday, this looks good from my perspective. Thanks, @mbondyra!

mbondyra commented:
_What should happen for array values and last_value?_ - I've changed the behavior to instead take the last value from the array.

mbondyra commented:
@elasticmachine merge upstream

stratoula reviewed (approved):
This works fine Marta and the bug is solved! I am approving. The only thing that I found is that when I have a gauge and open the dimensions editor, the chart jumps
![Lens-Elastic-_3_](https://user-images.githubusercontent.com/17003240/144841239-1af00f3d-53ca-44d0-a60b-ee4ca5f3fad8.gif)

I am not sure if it is caused by this PR, I can't replicate in master though

mbondyra commented:
Thanks Stratoula, I'll css things around to see if it can help!
